File: Chapter02/VB/VBProject_02/ExtendControl/3-CustomControl/SlideshowCustomControl/Slideshow.vb
Imports System
Imports System.Diagnostics
Imports System.Text.RegularExpressions
Imports System.Windows
Imports System.Windows.Controls
Imports System.Windows.Markup
Imports System.Windows.Media.Imaging
Imports System.Collections.Generic

' Namespace SlideshowCustomControl

    Public class Slideshow
        Inherits Control

        Private layoutRoot As Canvas
        Private SlideImage As Image
        Private SlideBorder As Border
        Private SlideGrid As Grid
        Private SlideTitle As TextBlock
        Private LeftButton As Image
        Private RightButton As Image
        Private FullscreenButton As Image

    Public Sub New()
        MyBase.New()

        DefaultStyleKey = GetType(Slideshow)

        AddHandler Application.Current.Host.Content.Resized, AddressOf Content_Resized
    End Sub '   New

    Public ReadOnly SlideStyleProperty As DependencyProperty = DependencyProperty.Register("SlideStyle", GetType(Style), GetType(Slideshow), Nothing)
    Public ReadOnly TextProperty As DependencyProperty = DependencyProperty.Register("Text", GetType(String), GetType(Slideshow), Nothing)
    Public ReadOnly SlideWidthProperty As DependencyProperty = DependencyProperty.Register("SlideWidth", GetType(Double), GetType(Slideshow), Nothing)
    Public ReadOnly SlideHeightProperty As DependencyProperty = DependencyProperty.Register("SlideHeight", GetType(Double), GetType(Slideshow), Nothing)
    Public ReadOnly MainTitleProperty As DependencyProperty = DependencyProperty.Register("MainTitle", GetType(String), GetType(Slideshow), Nothing)
    Public ReadOnly MainImageProperty As DependencyProperty = DependencyProperty.Register("MainImage", GetType(String), GetType(Slideshow), Nothing)
    Public ReadOnly CurrentSlideProperty As DependencyProperty = DependencyProperty.Register("CurrentSlide", GetType(Integer), GetType(Slideshow), Nothing)
    Public ReadOnly SlideCountProperty As DependencyProperty = DependencyProperty.Register("SlideCount", GetType(Integer), GetType(Slideshow), Nothing)
     ''' <summary>
     ''' Gets or sets the style of the added TextBlock controls
     ''' </summary>
    Public Property SlideStyle() As Style
        Get
            Return CType(Me.GetValue(SlideStyleProperty), Style)
        End Get

        Set(value As Style)
            MyBase.SetValue(SlideStyleProperty, CType(value, DependencyObject))
        End Set
    End Property
    ''' <summary>
    ''' Gets or sets the text of the LinkLabel control
    ''' </summary>
    Public Property Text() As String
        Get
            Return CType(Me.GetValue(TextProperty), String)
        End Get

        Set(value As String)
            MyBase.SetValue(TextProperty, value)
        End Set
    End Property

    Public Property SlideCount() As Integer
        Get
            Return CType(GetValue(SlideCountProperty), Integer)
        End Get

        Set(value As Integer)
            SetValue(SlideCountProperty, value)
        End Set
    End Property

    Public Property CurrentSlide() As Integer
        Get
            Return CType(GetValue(CurrentSlideProperty), Integer)
        End Get

        Set(value As Integer)
            SetValue(CurrentSlideProperty, value)
        End Set
    End Property

    Public Property SlideWidth() As Double
        Get
            Return CType(GetValue(SlideWidthProperty), Double)
        End Get

        Set(value As Double)
            SetValue(SlideWidthProperty, value)
        End Set
    End Property

    Public Property SlideHeight() As Double
        Get
            Return CType(GetValue(SlideHeightProperty), Double)
        End Get
        Set(value As Double)
            SetValue(SlideHeightProperty, value)
        End Set
    End Property

    Public Property MainImage() As String
        Get
            Return CType(GetValue(MainImageProperty), String)
        End Get

        Set(value As String)
            SetValue(MainImageProperty, value)
        End Set
    End Property

    Public Property MainTitle() As String
        Get
            Return CType(GetValue(MainTitleProperty), String)
        End Get

        Set(value As String)
            SetValue(MainTitleProperty, value)
        End Set
    End Property

    Public Overrides Sub OnApplyTemplate()

        Me.layoutRoot = CType(Me.GetTemplateChild("LayoutRoot"), Canvas)
        Me.SlideImage = CType(Me.GetTemplateChild("SlideImage"), Image)
        Me.SlideGrid = CType(Me.GetTemplateChild("SlideGrid"), Grid)
        Me.SlideBorder = CType(Me.GetTemplateChild("SlideBorder"), Border)
        Me.SlideTitle = CType(Me.GetTemplateChild("SlideTitle"), TextBlock)
        Me.LeftButton = CType(Me.GetTemplateChild("LeftButton"), Image)
        Me.RightButton = CType(Me.GetTemplateChild("RightButton"), Image)
        Me.FullscreenButton = CType(Me.GetTemplateChild("FullscreenButton"), Image)

        AddHandler SlideImage.MouseLeftButtonDown, AddressOf Me.NextSlide
        AddHandler SlideImage.ImageFailed, AddressOf Me.ImageFailed
        AddHandler RightButton.MouseLeftButtonDown, AddressOf Me.NextSlide
        AddHandler LeftButton.MouseLeftButtonDown, AddressOf Me.PreviousSlide
        AddHandler FullscreenButton.MouseLeftButtonUp, AddressOf Me.FullScreen

        MyBase.OnApplyTemplate()
        LoadItems()
    End Sub '   OnApplyTemplate

    Private Sub Content_Resized(sender As Object, e As Eventargs)

        Try
            If (Application.Current.Host.Content.IsFullScreen) Then

                SlideWidth = Application.Current.Host.Content.ActualWidth
                SlideHeight = Application.Current.Host.Content.ActualHeight
            Else
                SlideWidth = 400
                SlideHeight = 300
            End If

            Me.SlideImage.Width = SlideWidth
            Me.SlideGrid.Width = SlideWidth
            Me.SlideBorder.Width = SlideWidth

            Me.SlideImage.Height = SlideHeight - 35
            Me.SlideGrid.Height = SlideHeight
            Me.SlideBorder.Height = SlideHeight
        Catch ex As Exception
        End Try
    End Sub '   Content_Resized

    Private Sub FullScreen(sender As Object, e As Eventargs)

        Application.Current.Host.Content.IsFullScreen = Not Application.Current.Host.Content.IsFullScreen

        If (Application.Current.Host.Content.IsFullScreen) Then
            SlideWidth = Application.Current.Host.Content.ActualWidth
            SlideHeight = Application.Current.Host.Content.ActualHeight
        Else
            SlideWidth = 400
            SlideHeight = 300
        End If

        Me.SlideImage.Width = SlideWidth
        Me.SlideGrid.Width = SlideWidth
        Me.SlideBorder.Width = SlideWidth

        Me.SlideImage.Height = SlideHeight - 35
        Me.SlideGrid.Height = SlideHeight
        Me.SlideBorder.Height = SlideHeight
    End Sub '   FullScreen

    Private Sub ImageFailed(sender As Object, e As Eventargs)

        MainImage = "Images/_ErrorImage.jpg"
        SlideImage.Source = New BitmapImage(New Uri(MainImage, UriKind.Relative))
    End Sub '   ImageFailed

    Private Sub NextSlide(sender As Object, e As Routedeventargs)

        If (CurrentSlide = SlideCount - 1) Then
            CurrentSlide = 0
        Else
            CurrentSlide += 1
        End If

        Dim mySlides As SlideImagesList = CType(DataContext, SlideImagesList)

        MainImage = mySlides(CurrentSlide).ImageUri

        Dim CurrentSlideIndexed As Integer = CurrentSlide + 1

        MainTitle = CStr(CurrentSlideIndexed) + " / " + CStr(SlideCount)
        SlideImage.Source = New BitmapImage(New Uri(MainImage, UriKind.Relative))
        SlideTitle.Text = MainTitle
    End Sub '   NextSlide

    Private Sub PreviousSlide(sender As Object, e As Routedeventargs)

        If (CurrentSlide = 0) Then
            CurrentSlide = SlideCount - 1
        Else
            CurrentSlide -= 1
        End If

        Dim mySlides As SlideImagesList = CType(DataContext, SlideImagesList)

        MainImage = mySlides(CurrentSlide).ImageUri

        Dim CurrentSlideIndexed As Integer = CurrentSlide + 1

        MainTitle = CStr(CurrentSlideIndexed) + " / " + CStr(SlideCount)
        SlideImage.Source = New BitmapImage(New Uri(MainImage, UriKind.Relative))
        SlideTitle.Text = MainTitle
    End Sub '   PreviousSlide

    Private Sub LoadItems()

        Dim mySlides As SlideImagesList = CType(DataContext, SlideImagesList)

        SlideCount = mySlides.Count
        SlideImage.Source = New BitmapImage(New Uri(MainImage, UriKind.Relative))
        SlideTitle.Text = MainTitle

        Me.SlideImage.Width = SlideWidth
        Me.SlideGrid.Width = SlideWidth
        Me.SlideBorder.Width = SlideWidth

        Me.SlideImage.Height = SlideHeight - 35
        Me.SlideGrid.Height = SlideHeight
        Me.SlideBorder.Height = SlideHeight
    End Sub '   LoadItems
End Class   '   Slideshow

Public Class Slides

    Public Property ImageUri As String
    Public Property Title As String
End Class   '   Slides

Public Class SlideImagesList
    Inherits List(Of Slides)

    Dim si As Slides

    Public Property Val() As Slides
        Get
            Return si
        End Get
        Set(value As Slides)
            si = value
        End Set
    End Property

    Public Sub New()
    End Sub '   New
End Class   '   SlideImagesList
' End Namespace
' ..\ExtendControl\E\ExtendControl\ExtendControl\3-CustomControl\SlideshowCustomControl\Slideshow.cs

File: Chapter02/C#/Project/ExtendControl/3-CustomControl/SlideshowCustomControl/Slideshow.cs
using System;
using System.Diagnostics;
using System.Text.RegularExpressions;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Markup;
using System.Windows.Media.Imaging;
using System.Collections.Generic;

namespace SlideshowCustomControl
{
    public class Slideshow : Control
    {
        private Canvas layoutRoot;
        private Image SlideImage;
        private Border SlideBorder;
        private Grid SlideGrid;
        private TextBlock SlideTitle;
        private Image LeftButton;
        private Image RightButton;
        private Image FullscreenButton;

         public Slideshow() : base()
          {
              DefaultStyleKey = typeof(Slideshow);
              Application.Current.Host.Content.Resized += new EventHandler(Content_Resized);
          }
     public static readonly DependencyProperty SlideStyleProperty = DependencyProperty.Register("SlideStyle", typeof(Style), typeof(Slideshow), null);
     public static readonly DependencyProperty TextProperty = DependencyProperty.Register("Text", typeof(string), typeof(Slideshow), null);
     public static readonly DependencyProperty SlideWidthProperty = DependencyProperty.Register("SlideWidth", typeof(double), typeof(Slideshow), null);
     public static readonly DependencyProperty SlideHeightProperty = DependencyProperty.Register("SlideHeight", typeof(double), typeof(Slideshow), null);
     public static readonly DependencyProperty MainTitleProperty = DependencyProperty.Register("MainTitle", typeof(string), typeof(Slideshow), null);
     public static readonly DependencyProperty MainImageProperty = DependencyProperty.Register("MainImage", typeof(string), typeof(Slideshow), null);
     public static readonly DependencyProperty CurrentSlideProperty = DependencyProperty.Register("CurrentSlide", typeof(int), typeof(Slideshow), null);
     public static readonly DependencyProperty SlideCountProperty = DependencyProperty.Register("SlideCount", typeof(int), typeof(Slideshow), null);
     /// <summary>
     /// Gets or sets the style of the added TextBlock controls
     /// </summary>
     public Style SlideStyle
     {
         get
         {
             return (Style)this.GetValue(SlideStyleProperty);
         }

         set
         {
             base.SetValue(SlideStyleProperty, (DependencyObject)value);
         }
     }
     /// <summary>
     /// Gets or sets the text of the LinkLabel control
     /// </summary>
     public string Text
     {
         get
         {
             return (string)this.GetValue(TextProperty);
         }

         set
         {
             base.SetValue(TextProperty, value);
         }
     }
     public int SlideCount
     {
         get
         {
             return (int)GetValue(SlideCountProperty);
         }

         set
         {
             SetValue(SlideCountProperty, value);
         }

     }
     public int CurrentSlide
     {
         get
         {
             return (int)GetValue(CurrentSlideProperty);
         }

         set
         {
             SetValue(CurrentSlideProperty, value);
         }
     }
     public Double SlideWidth
     {
         get
         {
             return (double)GetValue(SlideWidthProperty);
         }

         set
         {
             SetValue(SlideWidthProperty, value);
           
         }
     }
     public Double SlideHeight
     {
         get
         {
             return (double)GetValue(SlideHeightProperty);
         }
         set
         {
             SetValue(SlideHeightProperty, value);
         }
     }
     public string MainImage
     {
         get
         {
             return (string)GetValue(MainImageProperty);
         }

         set
         {
             SetValue(MainImageProperty, value);
         }
     }
     public string MainTitle
     {
         get
         {
             return (string)GetValue(MainTitleProperty);
         }

         set
         {
             SetValue(MainTitleProperty, value);
         }
     }
     public override void OnApplyTemplate()
        {
           this.layoutRoot = this.GetTemplateChild("LayoutRoot") as Canvas;
           this.SlideImage = this.GetTemplateChild("SlideImage") as Image;
           this.SlideGrid = this.GetTemplateChild("SlideGrid") as Grid;
           this.SlideBorder = this.GetTemplateChild("SlideBorder") as Border;
           this.SlideTitle = this.GetTemplateChild("SlideTitle") as TextBlock;
           this.LeftButton = this.GetTemplateChild("LeftButton") as Image;
           this.RightButton = this.GetTemplateChild("RightButton") as Image;
           this.FullscreenButton = this.GetTemplateChild("FullscreenButton") as Image;

           this.SlideImage.MouseLeftButtonDown += new System.Windows.Input.MouseButtonEventHandler(this.NextSlide);
           this.SlideImage.ImageFailed += new EventHandler<ExceptionRoutedEventArgs> (this.ImageFailed);         
           this.RightButton.MouseLeftButtonDown += new System.Windows.Input.MouseButtonEventHandler(this.NextSlide);
           this.LeftButton.MouseLeftButtonDown += new System.Windows.Input.MouseButtonEventHandler(this.PreviousSlide);
           this.FullscreenButton.MouseLeftButtonUp += new System.Windows.Input.MouseButtonEventHandler(this.FullScreen);

           base.OnApplyTemplate();
           LoadItems();
        }
     private void Content_Resized(object sender, EventArgs e)
     {
         try
         {
             if (Application.Current.Host.Content.IsFullScreen)
             {

                 SlideWidth = Application.Current.Host.Content.ActualWidth;
                 SlideHeight = Application.Current.Host.Content.ActualHeight;

             }
             else
             {
                 SlideWidth = 400;
                 SlideHeight = 300;
             }
             this.SlideImage.Width = SlideWidth;
             this.SlideGrid.Width = SlideWidth;
             this.SlideBorder.Width = SlideWidth;

             this.SlideImage.Height = SlideHeight - 35;
             this.SlideGrid.Height = SlideHeight;
             this.SlideBorder.Height = SlideHeight;
         }
         catch (Exception)
         { }
     }
     private void FullScreen(object sender, EventArgs e)
     {
         Application.Current.Host.Content.IsFullScreen = !Application.Current.Host.Content.IsFullScreen;
         if (Application.Current.Host.Content.IsFullScreen)
         {
             SlideWidth = Application.Current.Host.Content.ActualWidth;
             SlideHeight = Application.Current.Host.Content.ActualHeight;
         }
         else
         {
             SlideWidth = 400;
             SlideHeight = 300;

         }
         this.SlideImage.Width = SlideWidth;
         this.SlideGrid.Width = SlideWidth;
         this.SlideBorder.Width = SlideWidth;

         this.SlideImage.Height = SlideHeight - 35;
         this.SlideGrid.Height = SlideHeight;
         this.SlideBorder.Height = SlideHeight;
     }
     private void ImageFailed(object sender, EventArgs e)
     {
         MainImage = "Images/_ErrorImage.jpg";
         SlideImage.Source = new BitmapImage(new Uri(MainImage, UriKind.Relative));
     }
     private void NextSlide(object sender, RoutedEventArgs e)
     {
         if (CurrentSlide == SlideCount - 1)
             CurrentSlide = 0;
         else
             CurrentSlide++;
         SlideImagesList mySlides = DataContext as SlideImagesList;
         MainImage = mySlides[CurrentSlide].ImageUri;
         int CurrentSlideIndexed = CurrentSlide + 1;
         MainTitle = CurrentSlideIndexed + " / " + SlideCount;
         SlideImage.Source = new BitmapImage(new Uri(MainImage, UriKind.Relative));
         SlideTitle.Text = MainTitle;
     }

     private void PreviousSlide(object sender, RoutedEventArgs e)
     {
         if (CurrentSlide == 0)
             CurrentSlide = SlideCount - 1;
         else
             CurrentSlide--;
         SlideImagesList mySlides = DataContext as SlideImagesList;
         MainImage = mySlides[CurrentSlide].ImageUri;
         int CurrentSlideIndexed = CurrentSlide + 1;
         MainTitle = CurrentSlideIndexed + " / " + SlideCount;
         SlideImage.Source = new BitmapImage(new Uri(MainImage, UriKind.Relative));
         SlideTitle.Text = MainTitle;
     }
     private void LoadItems()
     {
         SlideImagesList mySlides = DataContext as SlideImagesList;
         SlideCount = mySlides.Count;
         SlideImage.Source = new BitmapImage(new Uri(MainImage, UriKind.Relative));
         SlideTitle.Text = MainTitle;

         this.SlideImage.Width = SlideWidth;
         this.SlideGrid.Width = SlideWidth;
         this.SlideBorder.Width = SlideWidth;

         this.SlideImage.Height = SlideHeight- 35;
         this.SlideGrid.Height = SlideHeight;
         this.SlideBorder.Height = SlideHeight;
     }
   }
    public class Slides
    {
        public string ImageUri { get; set; }
        public string Title { get; set; }
    }
    public class SlideImagesList : List<Slides>
    {
        Slides si;
        public Slides Val { get { return si; } set { si = value; } }
        public SlideImagesList()
        {
        }
    }
}


